### PR TITLE
[wasm] [llvm] Support wasm with single thread

### PR DESCRIPTION
**Pull request by squarefk**
Related issue = [LAN-43]

[LAN-43]: https://taichi-graphics.atlassian.net/browse/LAN-43

### CODE REVIEW

File: taichi/backends/cpu/codegen_cpu.cpp [hunk 1]
```diff
@@ -29,32 +29,88 @@ class CodeGenLLVMCPU : public CodeGenLLVM {
       step = -1;
     }
 
-    auto *tls_prologue = create_xlogue(stmt->tls_prologue);
```
Comment (k-ye) [inline taichi/backends/cpu/codegen_cpu.cpp:32]: nit: Create a new `codegen_wasm.cpp`

File: taichi/backends/cpu/codegen_cpu.cpp [hunk 2]
```diff
@@ -29,32 +29,88 @@ class CodeGenLLVMCPU : public CodeGenLLVM {
...
-
-    // The loop body
-    llvm::Function *body;
-    {
-      auto guard = get_function_creation_guard(
-          {llvm::PointerType::get(get_runtime_type("Context"), 0),
-           llvm::Type::getInt8PtrTy(*llvm_context),
-           tlctx->get_data_type<int>()});
+    if (prog->config.arch == Arch::wasm) {
+      using namespace llvm;
```
Comment (k-ye) [inline taichi/backends/cpu/codegen_cpu.cpp:33]: Remove `using namespace llvm`. Let's use `auto*` if it's redundant to type `llvm::BasicBlock` twice.

File: taichi/backends/wasm/codegen_wasm.cpp [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    if (prog->config.kernel_profiler && arch_is_cpu(prog->config.arch)) {
+      call(builder.get(), "LLVMRuntime_profiler_stop", {get_runtime()});
+    }
+    finalize_offloaded_task_function();
+    current_task->end();
+    current_task = nullptr;
+    current_offload = nullptr;
+  }
+
+  void visit(ExternalFuncCallStmt *stmt) override {
```
Comment (k-ye) [inline taichi/backends/wasm/codegen_wasm.cpp:122]: Remove `void visit(ExternalFuncCallStmt *stmt) override` function

File: taichi/backends/wasm/codegen_wasm.cpp [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    if (prog->config.kernel_profiler && arch_is_cpu(prog->config.arch)) {
+      call(
+          builder.get(), "LLVMRuntime_profiler_start",
+          {get_runtime(), builder->CreateGlobalStringPtr(offloaded_task_name)});
+    }
+    if (stmt->task_type == Type::serial) {
+      stmt->body->accept(this);
+    } else if (stmt->task_type == Type::range_for) {
+      create_offload_range_for(stmt);
+    } else if (stmt->task_type == Type::struct_for) {
```
Comment (k-ye) [inline taichi/backends/wasm/codegen_wasm.cpp:102]: Remove L102-115

File: taichi/backends/wasm/codegen_wasm.cpp [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    builder->CreateCall(func, arg_values);
+  }
+};
+
+FunctionType CodeGenWASM::codegen() {
+  TI_AUTO_PROF
+  return CodeGenLLVMWASM(kernel, ir).gen();
+}
+
+TLANG_NAMESPACE_END
```
Comment (k-ye) [inline taichi/backends/wasm/codegen_wasm.cpp:155]:
```suggestion
}
}
```

File: taichi/backends/wasm/codegen_wasm.h [hunk 6]
```diff
@@ -0,0 +1,17 @@
+// x86 backend implementation
+
+#pragma once
+
+#include "taichi/codegen/codegen.h"
+
+TLANG_NAMESPACE_BEGIN
```
Comment (k-ye) [inline taichi/backends/wasm/codegen_wasm.h:7]:
```suggestion
namespace taichi {
namespace lang {
```

File: taichi/backends/wasm/codegen_wasm.h [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+class CodeGenWASM : public KernelCodeGen {
+ public:
+  CodeGenWASM(Kernel *kernel, IRNode *ir = nullptr) : KernelCodeGen(kernel, ir) {
+  }
+
+  virtual FunctionType codegen() override;
+};
+
+TLANG_NAMESPACE_END
```
Comment (k-ye) [inline taichi/backends/wasm/codegen_wasm.h:17]:
```suggestion
}
}
```

File: taichi/backends/wasm/codegen_wasm.h (hunk 6 above)
Comment (squarefk) [reply taichi/backends/wasm/codegen_wasm.h]: Done

File: taichi/backends/wasm/codegen_wasm.h (hunk 7 above)
Comment (squarefk) [reply taichi/backends/wasm/codegen_wasm.h]: Done

File: taichi/backends/wasm/codegen_wasm.cpp (hunk 5 above)
Comment (squarefk) [reply taichi/backends/wasm/codegen_wasm.cpp]: Done

File: taichi/backends/wasm/codegen_wasm.cpp (hunk 4 above)
Comment (squarefk) [reply taichi/backends/wasm/codegen_wasm.cpp]: Leave 'else' branch with TI_NOT_IMPLEMENTED
Done

File: taichi/backends/wasm/codegen_wasm.cpp (hunk 3 above)
Comment (squarefk) [reply taichi/backends/wasm/codegen_wasm.cpp]: Done

File: taichi/backends/wasm/codegen_wasm.cpp [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+#include "taichi/codegen/codegen_llvm.h"
+#include "taichi/common/core.h"
+#include "taichi/util/io.h"
+#include "taichi/lang_util.h"
+#include "taichi/program/program.h"
+#include "taichi/ir/ir.h"
+#include "taichi/ir/statements.h"
+#include "taichi/util/statistics.h"
+
+TLANG_NAMESPACE_BEGIN
```
[the remark this answers is not in the document]
Comment (squarefk) [reply taichi/backends/wasm/codegen_wasm.cpp]: Done

File: taichi/backends/wasm/codegen_wasm.cpp [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    int step = 1;
+
+    // In parallel for-loops reversing the order doesn't make sense.
+    // However, we may need to support serial offloaded range for's in the
+    // future, so it still makes sense to reverse the order here.
+    if (stmt->reversed) {
+      step = -1;
+    }
+
+    using namespace llvm;
```
[the remark this answers is not in the document]
Comment (squarefk) [reply taichi/backends/wasm/codegen_wasm.cpp]: Done

File: taichi/backends/cpu/codegen_cpu.cpp (hunk 2 above)
Comment (squarefk) [reply taichi/backends/cpu/codegen_cpu.cpp]: Done

File: taichi/backends/cpu/codegen_cpu.cpp (hunk 1 above)
Comment (squarefk) [reply taichi/backends/cpu/codegen_cpu.cpp]: Done

**Review by k-ye — APPROVED** [review]
LGTM thanks!

**Comment by k-ye** [comment]
Test [LAN-43]

**Pull request merged**